File: src/agent/detect.ts
import type { BackendProbe, PermissionMode } from './types';
import { DEFAULT_BACKEND_ID } from './types';
import { BACKEND_CATALOG, catalogById, type AgentFamily } from './catalog';
import { resolveCodexBin, codexVersionAsync } from './codex-appserver/locate';

/**
 * 按 agent 维度检测（backend-detection.md §1）：探底层 agent（当前仅 codex）→ 推导
 * 每个后端的可用性，并据此算「有效默认后端」。后端只是 agent 的「接入方式」——本机
 * 能力面按 agent 组织，与运行时工厂表（REGISTRY）解耦，便于将来再挂新 agent。
 *
 * 智能默认：有 codex → codex-appserver；否则回退 codex-appserver 占位（doctor 会报
 * 需安装）。
 */

export type AgentId = 'codex' | (string & {});

/** 一个后端在本机的可用性（按 agent 推导，挂 catalog 元数据）。 */
export interface BackendAvailability {
  backendId: string;
  available: boolean;
  /** 不可用原因（探测 hint / 推导出的缺失项）。 */
  reason?: string;
  version: string | null;
  supportedModes?: readonly PermissionMode[];
  /** 可一键按需下载（npm-ondemand 且当前未装）。 */
  installable: boolean;
}

/** 一个底层 agent 的运行时探测结果 + 它衍生的后端可用性。 */
export interface AgentRuntime {
  id: AgentId;
  displayName: string;
  installed: boolean;
  version: string | null;
  backends: BackendAvailability[];
  /** installed=false 时的安装提示。 */
  installHint?: string;
}

/** 探 codex agent（复用 resolveCodexBin + codexVersionAsync）。 */
async function probeCodexAgent(): Promise<AgentRuntime> {
  const entry = BACKEND_CATALOG.find((e) => e.id === 'codex-appserver')!;
  const bin = resolveCodexBin({ force: true });
  const version = bin ? await codexVersionAsync(bin, { force: true }) : null;
  const installed = !!bin && !!version;
  return {
    id: 'codex',
    displayName: 'Codex',
    installed,
    version,
    installHint: installed ? undefined : entry.dep.detectHint,
    backends: [
      {
        backendId: 'codex-appserver',
        available: installed,
        reason: !bin ? entry.dep.detectHint : !version ? 'codex --version 失败' : undefined,
        version,
        supportedModes: entry.supportedModes,
        installable: false,
      },
    ],
  };
}

/** 探全部 agent（当前仅 codex），并行。绝不抛错（各 probe 自身降级）。 */
export async function detectAgents(): Promise<AgentRuntime[]> {
  return Promise.all([probeCodexAgent()]);
}

/** 从一次 detectAgents 结果挑默认后端（智能默认规则）。 */
export function pickDefaultBackend(agents: AgentRuntime[]): string {
  const find = (id: string): BackendAvailability | undefined =>
    agents.flatMap((a) => a.backends).find((b) => b.backendId === id);
  const pickable = (id: string) => !catalogById(id)?.hidden && find(id)?.available;
  if (pickable('codex-appserver')) return 'codex-appserver';
  return DEFAULT_BACKEND_ID; // 无 codex → codex 占位（doctor 会报需安装）
}

/**
 * 有效默认后端：缓存一次 detectAgents → pickDefaultBackend。散落的
 * `?? DEFAULT_BACKEND_ID` 应收敛到这里 / {@link backendForProject}，避免硬编码
 * 常量回退绕过智能默认。探测失败回退 DEFAULT_BACKEND_ID（保持现状语义）。
 *
 * 缓存：detect 要 spawn（codex --version），daemon 生命周期内本机装配几乎不变；
 * `force` 绕过缓存（体检 / 用户刚装完后要看「现在」的默认）。
 */
let defaultCache: string | undefined;
export async function effectiveDefaultBackend(opts?: { force?: boolean }): Promise<string> {
  if (!opts?.force && defaultCache !== undefined) return defaultCache;
  try {
    defaultCache = pickDefaultBackend(await detectAgents());
  } catch {
    defaultCache = DEFAULT_BACKEND_ID;
  }
  return defaultCache;
}

/**
 * 一个项目实际用哪个后端：显式选择优先，否则用有效默认（智能默认）。
 * 收敛全部 `p.backend ?? DEFAULT_BACKEND_ID` 的回退点 —— 未显式设的项目运行时
 * 按「有效默认」路由（design backend-detection.md §3.2 方案B：全局默认 + 项目继承）。
 */
export async function backendForProject(
  p: { backend?: string },
  opts?: { force?: boolean },
): Promise<string> {
  // 显式选择优先——但 id 必须仍是已注册后端；指向已移除后端（如历史 claude-*）的旧
  // 配置回退到有效默认，避免 createBackend 抛「未知后端」。
  if (p.backend && catalogById(p.backend)) return p.backend;
  return effectiveDefaultBackend(opts);
}

/** 一个 agent 家族下的全部后端可用性（picker 分组用）。 */
export function familyOf(agents: AgentRuntime[], family: AgentFamily): AgentRuntime | undefined {
  // family 与 AgentId 在本设计里同名（如 codex），直接按 id 匹配。
  return agents.find((a) => a.id === family);
}

/** 一个后端探测态归一成 BackendProbe（detect → probe 适配，供 catalog/doctor 复用）。 */
export function availabilityToProbe(a: BackendAvailability): BackendProbe {
  return {
    ok: a.available,
    version: a.version,
    hint: a.available ? undefined : a.reason,
    installable: a.installable || undefined,
    depState: a.available ? 'installed' : a.installable ? 'not-installed' : 'external-missing',
  };
}

File: internal/agent/detect.go
package agent

// detect.go —— 后端探测 + 智能默认（对齐 TS agent/detect）。
// DetectAgents（探 codex）→ PickDefaultBackend → EffectiveDefaultBackend（缓存）→ BackendForProject。
// 依赖 catalog + codex.ResolveCodexBin/CodexVersion（通过 import codex 触发注册）。

import "sync"

// AgentID 底层 agent id（codex / 未来扩展）。
type AgentID = string

// BackendAvailability 一个后端在本机的可用性。
type BackendAvailability struct {
	BackendID      string
	Available      bool
	Reason         string // 不可用原因
	Version        string
	SupportedModes []PermissionMode
	Installable    bool
}

// AgentRuntime 一个底层 agent 的探测结果 + 衍生后端可用性。
type AgentRuntime struct {
	ID          AgentID
	DisplayName string
	Installed   bool
	Version     string
	Backends    []BackendAvailability
	InstallHint string
}

// ProbeFunc 探测单个后端（注入，避免 agent 包依赖 codex 子包）。
type ProbeFunc func() (bin string, version string, ok bool, hint string)

// DetectAgentsFns 可注入的 detect 实现列表（各后端包 init 通过 RegisterDetect 追加）。
// 单变量会被后初始化的包覆盖，故用 slice 累积，支持同时探测多个后端（如 codex + claude）。
var DetectAgentsFns []func() []AgentRuntime

// RegisterDetect 追加一个后端探测函数（幂等由调用方保证）。
func RegisterDetect(fn func() []AgentRuntime) {
	DetectAgentsFns = append(DetectAgentsFns, fn)
}

// DetectAgents 探全部已注册 agent（codex / claude 等，经各自 init 注入）。
func DetectAgents() []AgentRuntime {
	out := []AgentRuntime{}
	for _, fn := range DetectAgentsFns {
		out = append(out, fn()...)
	}
	return out
}

// PickDefaultBackend 从 detectAgents 结果挑默认（有 codex→codex；无→占位）。
func PickDefaultBackend(agents []AgentRuntime) string {
	find := func(id string) *BackendAvailability {
		for _, a := range agents {
			for i := range a.Backends {
				if a.Backends[i].BackendID == id {
					return &a.Backends[i]
				}
			}
		}
		return nil
	}
	pickable := func(id string) bool {
		entry, ok := CatalogByID(id)
		if !ok || entry.Hidden {
			return false
		}
		ba := find(id)
		return ba != nil && ba.Available
	}
	if pickable(DEFAULT_BACKEND_ID) {
		return DEFAULT_BACKEND_ID
	}
	// codex 不可用 → 退到任一其它可用后端（如 claude），让只有 claude 的机器也能跑。
	for _, e := range BackendCatalog {
		if e.ID != DEFAULT_BACKEND_ID && pickable(e.ID) {
			return e.ID
		}
	}
	return DEFAULT_BACKEND_ID // 无可用后端 → codex 占位（doctor 报需装）
}

var (
	defaultCacheMu sync.Mutex
	defaultCache   string
	defaultCached  bool
)

// EffectiveDefaultBackend 有效默认（缓存 + force 绕过）。
func EffectiveDefaultBackend(force bool) string {
	defaultCacheMu.Lock()
	defer defaultCacheMu.Unlock()
	if !force && defaultCached {
		return defaultCache
	}
	agents := DetectAgents()
	defaultCache = PickDefaultBackend(agents)
	defaultCached = true
	return defaultCache
}

// BackendForProject 项目实际后端：显式选择优先（必须 catalog 注册），否则有效默认。
func BackendForProject(backend string, force bool) string {
	if backend != "" {
		if _, ok := CatalogByID(backend); ok {
			return backend
		}
	}
	return EffectiveDefaultBackend(force)
}

// AvailabilityToProbe 后端可用性 → BackendProbe 归一。
func AvailabilityToProbe(a BackendAvailability) BackendProbe {
	depState := "external-missing"
	if a.Available {
		depState = "installed"
	} else if a.Installable {
		depState = "not-installed"
	}
	hint := ""
	if !a.Available {
		hint = a.Reason
	}
	installable := a.Installable
	return BackendProbe{
		Ok: a.Available, Version: a.Version, Hint: hint,
		Installable: installable, DepState: depState,
	}
}

// FamilyOf 按 agent family 取 runtime（family 与 agent id 同名）。
func FamilyOf(agents []AgentRuntime, family AgentFamily) *AgentRuntime {
	for i := range agents {
		if agents[i].ID == string(family) {
			return &agents[i]
		}
	}
	return nil
}
